Fixes read_be and read_le to remove aliasing issue

joekale opened the pull request:
Also adds .vscode to .gitignore for VS Code users.

gtests confirm functionality with float, double, and uint32_t

Commit:
Fixes read_be and read_le to remove aliasing issue

Also adds .vscode to gitignore for VS Code users.

Commit:
removes unused include from debugging

JWhitleyWork commented on test/network_interface_test.cpp:
```diff
@@ -18,9 +18,9 @@ TEST(TCPInterface, testReadBEDouble)
-    ASSERT_EQ(1234567891, AS::Network::read_be<uint32_t>(bytes, 4, 0));
+    uint32_t in = 0x34FB5E38; //1234567891 as  hex Int
+    uint8_t bytes[4] = { 0x34, 0xFB, 0x5E, 0x38};
+    ASSERT_EQ(888888888, AS::Network::read_be<uint32_t>(bytes, 4, 0));
```
You changed the test value but not the comment that describes the value. Not sure which one you meant to change but they should be consistent (probably better to make them consistent between BE and LE tests too).

joekale replied:
I did notice this one and I will change that. I was seeing odd behavior and wanted to make sure it was consistent between 2 different integers.

JWhitleyWork commented on include/network_interface/network_utils.h:
```diff
@@ -83,14 +87,17 @@ namespace Network
-		T retVal = (*(reinterpret_cast<T *>(&rcvData)) * (T)factor) - valueOffset;
+		
+		T retVal;
+		std::memcpy(&retVal, &rcvData, sizeof(T));
```
The problem with using std::memcpy here is machine endianness. Whether the data gets copied from the left or right side of rcvData depends on the endianness of the machine that is running the program and since we are using bitshifts above to store the data, the underlying memory storage of said data could be different depending on the machine's endianness. The safest and most portable way to do this is to shift the value back out of rcvData. However, since we would be shifting it into rcvData and right back out again, why don't we just drop rcvData and shift it straight into a retVal of type T? As long as (sizeof(T) == size) ( which should probably be checked for) then the endianness becomes irrelevant.

JWhitleyWork replied:
I thought about dropping the size parameter altogether from these functions but we need to guarantee that the size of the buffer array being passed in is known and matches the size of the primitive we're trying to cast to. For this reason, it probably makes sense for us to throw an exception if they don't match.

joekale replied:
The problem with the shifting into T is that you can't bit shift a float. I did try that but it wouldn't compile.

JWhitleyWork replied:
So, one possibility is to handle floats specially. If sizeof(T) == sizeof(float), then memcpy it to a uint32_t. If sizeof(T) == sizeof(double), then memcpy it to a uint64_t. We probably won't need to handle the case of sizeof(long double) (128 bits).

JWhitleyWork replied:
In this case, memcpy is OK because the underlying storage order will be the same for the float and the unsigned integer and we aren't copying a subset.

JWhitleyWork reviewed (changes requested):
See my code comments.

srustanAStuff commented:
Does endianess matter in this context?  Approving with the understanding that it works with our testing.

JWhitleyWork commented:
Yes, it does. Bitshifting works regardless of machine endianness because the machine code instructions generated by the compiler are different depending on the endianness of the machine. However, std::memcpy will copy data from sequential, physical memory locations. The order in which the primitive is stored in those memory locations is machine-dependent. Even though we shifted it in in "big-endian" order, the underlying storage could be different. That's why assuming to copy a subset of bytes out of a larger primitive is dangerous. Shifting is the portable way to do it.

JWhitleyWork commented:
Also, it works with our testing because we've only tested on x86/x64 machines (which store in little-endian format). ARM processors are _bi-endian_ - you can flip a hardware flag to change the endianness of the processor.

joekale reviewed (commented):
The best I can find to explain (granted its StackOverflow but the explanation seems to align with my understanding) is that the start address of the memory on the machine is the same whether it is little or big endian. memcpy does a blind copy of memory from the start address or the source for a specific length to the destination. The order those bytes are stored doesn't really matter since they will be stored the same way for both variables. https://stackoverflow.com/questions/48619160/c-memcpy-copies-bytes-with-little-endianness

Commit:
changes comment on readbeint test case

joekale commented:
Not sure why I wasn't paying attention and reviewed my own code but my comment above should have been my reply.

srustanAStuff commented:
So endianess in the context of memcpy doesn't matter.

JWhitleyWork commented:
Here's where I disagree with you. The reads/writes/storage all have to be consistent. They read in the same order as they write (not in inverse). Because of this, on little-endian systems, it would store 0x12345678 as:

12 34 56 78 (LSB gets stored at lowest memory address)

And the start address would point to the location of 78 because that is the first byte to be stored and will be the first bye to be read. This means that a memcpy of 2 bytes would result in an underlying storage of:

56 78 (78 gets written first)

Which would be read out as: 0x5678 in a little-endian system.

On a big-endian system, the value 0x12345678 would be stored as:

78 56 34 12 (12 gets written first)

And the starting memory address would be where 12 is stored because that's the first byte written and will be the first byte to be read. A memcpy of 2 bytes would start at 12 and result in storage of:

34 12

Which would be read out as 0x1234 in a big-endian system.

JWhitleyWork commented:
I agree with what you said - the bytes will be interpreted the same because both are stored on the same machine - the problem is with cutting out a subset of those bytes in memory. Depending on which order they're stored in, you'll either get some MSBs or some LSBs when you do a memcpy.

Commit:
Adds check for system endianess

This allows moving the source address start further into
the stored uint64_t to properly grab the least significant
bytes for putting into smaller containers.

joekale commented:
I pushed a commit that adds a check for system endianess to move the start address of the copied bytes if on a big endian system.

JWhitleyWork commented on include/network_interface/network_utils.h:
```diff
@@ -24,6 +24,16 @@ namespace Network
+
+		return big_int.c[0] == 1; 
+	}
+
```
Would you mind implementing it as:

```
bool is_big_endian()
{
    union {
        uint32_t i;
        char c[4];
    } bint = {0x01020304};

    return bint.c[0] == 1; 
}
```

This one doesn't require making a copy of the read value each time and guarantees that the value being checked won't have 0x01 in both the LSB and the MSB (which would defeat the version you have above). This is a C++ version of the one at [https://stackoverflow.com/questions/1001307/detecting-endianness-programmatically-in-a-c-program](https://stackoverflow.com/questions/1001307/detecting-endianness-programmatically-in-a-c-program).

joekale replied:
yep, good call, commit on its way

JWhitleyWork commented:
See my comment. Nice job on the BE offset, BTW. Took me a while to run that one through my head.

Commit:
removes copying input into system endianess check

joekale commented:
Thanks. I had to get my head wrapped around it as well. once we all finally got on the same page with the issue it made the fix simpler in my head.

Commit:
endianess check now returns bool
